pycbc/strain/strain.py

```diff
@@ -1245,6 +1245,37 @@ def verify_segment_options_multi_ifo(cls, opt, parser, ifos):
             required_opts_multi_ifo(opt, parser, ifo, cls.required_opts_list)
 
 
+def create_class_fft_for_cache(npoints_time, delta_t, dtype, ifft=False):
```

Comment:
Does this function belong here? It seems like a convenience method for FFTs in general no?

Reply:
@ahnitz @spxiwh I don't think it makes sense to move this into the FFT library proper, or if you do, you need to generalize it substantially.  For instance, this code assumes you are transforming between a time series and a frequency series; it doesn't consider the  array to array class at all.

Also, please think about the logic here.  It may be true in a particular use case that any FFT instance of a given size can be reused, but that doesn't have to be true generally. It really depends on a particular user to know when an FFT with a certain set of its arguments the same as another can actually be reused; what is the "right choice" here is not (for example) the right choice in the MatchedFilter class.

Reply:
@ahnitz I moved this into pycbc.fft.class_api. Is this a better place for it??

Reply:
@josh-willis Would you be okay with this living in the fft module if it allowed generic array <-> array R2C or C2R transforms (with a "NotImplementedError if C2C or similar is attempted)? Normalizations are handled outside of the class based API anyway, so I think the only real change to do that is to convert back to FSeries/TSeries within the libraries being changed. If not, any suggestions on where should I put this?

Regarding logic and use cases. There are two places this is being added (for the moment):

* The Live data handler. This is specific to PyCBC Live and doesn't get called from anywhere else.
* The high-pass filter in strain.py, which is used by *many* codes. In this case the old behaviour will be the default, and only by very specifically saying "use class-based cached FFTs" will the new behaviour be activated.

.... I think that autogating may also need this if being used in the full PyCBC Live analysis, but again I'd use the same behaviour as that used in strain.py.

Reply:
@spxiwh I guess so?  I think my point is that the function should live at no higher level of generality than its use demands.  For example, if the autogating, pycbc_live, and HPF are all similar enough use cases that it should live in `strain` and be called from the codes that need it, then I would prefer that.  Things in FFT need to be broadly supporting FFTs as we use them.

I would clarify in the documentation (docstrnig) that this will create aligned-memory, out-of-place FFTs, if you leave it in the `fft/class_api.py`, since those are restrictions from the most general thing that the class-based FFT supports (ideally that would even be in the function name, but that is plenty long enough already).

Is there are reason that you create and manage the cache yourself, using python dictionaries, rather than using `functools.lru_cache` as you did before? Though code provided by others is not foolproof, I feel better using a widely used cache that is core python, and which if it has bugs will be something that others will fix. In particular, I worry about things being thread-safe; there is an open issue to clarify that `functools.lru_cache` is thread-safe [here](https://github.com/python/cpython/issues/93179); I don't know what the answer is for what you have created.